libbindgen/src/ir/context.rs

```diff
@@ -1066,3 +1102,38 @@ impl<'ctx, 'gen> Iterator for WhitelistedItemsIter<'ctx, 'gen>
         Some(id)
     }
 }
+
+pub struct RelatedItemsIter<'ctx, 'gen>
+    where 'gen: 'ctx,
+{
+    ctx: &'ctx BindgenContext<'gen>,
+    seen: ItemSet,
+    to_iterate: VecDeque<ItemId>,
+}
+
+impl<'ctx, 'gen> Iterator for RelatedItemsIter<'ctx, 'gen>
+    where 'gen: 'ctx,
+{
+    type Item = ItemId;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let id = match self.to_iterate.pop_front() {
+            None => return None,
```

Comment:
Instead of eagerly returning `None`, we should see if there is a next root:

```rust
let id = match self.to_iterate.pop_front() {
    None => {
        // We've traversed everything reachable from the previous root(s), see if
        // we have any more roots.
        //
        // (Note: The `self.all_items` here would be the iterator returned by `ctx.items()`.)
        match self.all_items.filter(|&(id, _)| !self.seen.contains_key(id)).next() {
            None => return None,
            Some(id) => {
                 // This is a new root.
                 self.seen.insert(id, id);
                 id
            }
    },
    Some(id) => id,
};
```